File: backend/app/services/image_search_service.py
import os
import requests

SERPAPI_KEY = os.getenv("SERPAPI_KEY")
SERPAPI_URL = "https://serpapi.com/search"


def search_background_images(query: str, num_results: int = 12) -> list[dict]:
    """
    Searches Google Images via SerpAPI for background image candidates.
    Returns a list of dicts: {thumbnail, original, title, source}
    """
    if not SERPAPI_KEY:
        raise RuntimeError("SERPAPI_KEY is not set in environment")

    params = {
        "engine": "google_images",
        "q": query,
        "api_key": SERPAPI_KEY,
        "num": num_results,
    }

    response = requests.get(SERPAPI_URL, params=params, timeout=15)
    response.raise_for_status()
    data = response.json()

    images = data.get("images_results", [])[:num_results]

    results = []
    for img in images:
        results.append({
            "thumbnail": img.get("thumbnail"),
            "original": img.get("original"),
            "title": img.get("title"),
            "source": img.get("source"),
        })

    return results

File: frontend/lib/image-search-service.ts
interface ImageResult {
  thumbnail: string;
  original: string;
  title: string;
  source: string;
}

/**
 * Search for background images using Pixabay API
 * 
 * @param query Search query string
 * @param numResults Number of results to return (default: 12, minimum: 3 per Pixabay requirement)
 * @returns Promise<ImageResult[]> Array of image results
 * @throws Error if PIXABAY_API_KEY is missing or API call fails
 */
export async function searchBackgroundImages(
  query: string,
  numResults: number = 12
): Promise<ImageResult[]> {
  const apiKey = process.env.PIXABAY_API_KEY;
  
  if (!apiKey) {
    throw new Error("PIXABAY_API_KEY environment variable is not configured");
  }

  // Ensure minimum results per Pixabay requirement
  const perPage = Math.max(numResults, 3);

  // Build Pixabay API URL
  const url = new URL("https://pixabay.com/api/");
  url.searchParams.set("key", apiKey);
  url.searchParams.set("q", encodeURIComponent(query));
  url.searchParams.set("image_type", "photo");
  url.searchParams.set("per_page", perPage.toString());
  url.searchParams.set("safesearch", "true");
  url.searchParams.set("category", "backgrounds,places,nature,business,people");

  try {
    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Pixabay API request failed: ${response.status} ${response.statusText}`);
    }

    const data = await response.json();

    // Check for API errors
    if (data.error) {
      throw new Error(`Pixabay API error: ${data.error}`);
    }

    // Extract hits array
    const hits = data.hits || [];
    
    if (hits.length === 0) {
      return [];
    }

    // Transform Pixabay results to our format
    const results: ImageResult[] = hits
      .slice(0, numResults)
      .map((hit: any) => ({
        thumbnail: hit.webformatURL || "",
        original: hit.largeImageURL || hit.fullHDURL || hit.webformatURL || "",
        title: hit.tags || "",
        source: "Pixabay",
      }))
      // Filter out any results with missing essential fields
      .filter((item: ImageResult) => item.thumbnail && item.original);

    return results;
  } catch (error) {
    // Re-throw with more context
    if (error instanceof Error) {
      throw new Error(`Image search failed: ${error.message}`);
    }
    throw new Error("Image search failed: Unknown error");
  }
}